src/core/detail/btree_internal.h

```diff
@@ -0,0 +1,458 @@
+  }
+
+  Key_t Key(unsigned index) const {
+    Key_t res;
+    memcpy(&res, KeyPtr(index), kKeySize);
+    return res;
+  }
+
+  void SetKey(size_t index, Key_t item) {
+    uint8_t* slot = KeyPtr(index);
+    memcpy(slot, &item, kKeySize);
+  }
+
+  BTreeNode** Children() {
+    // aligned to 8 bytes.
+    constexpr uint16_t kChildOffset = kMaxInnerKeys * kKeySize + kKeyOffset;
+
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(this) + kChildOffset;
+    return reinterpret_cast<BTreeNode**>(ptr);
+  }
+
+  BTreeNode* Child(unsigned i) {
+    return Children()[i];
+  }
+
+  void SetChild(unsigned i, BTreeNode* child) {
+    Children()[i] = child;
+  }
+
+  struct SearchResult {
+    uint32_t index : 31;
+    uint32_t found : 1;
+  };
+
+  // Searches for key in the node using binary seach.
+  // Returns SearchResult with index of the key if found.
+  SearchResult BSearch(Key_t key) const;
+
+  void Split(BTreeNode* right, Key_t* median);
+
+  bool IsLeaf() const {
+    return leaf_;
+  }
+
+  unsigned NumItems() const {
+    return num_items_;
+  }
+
+  unsigned AvailableSlotCount() const {
+    return MaxItems() - num_items_;
+  }
+
+  unsigned MaxItems() const {
+    return leaf_ ? kMaxLeafKeys : kMaxInnerKeys;
+  }
+
+  unsigned MinItems() const {
+    return leaf_ ? kMinLeafKeys : kMinInnerKeys;
+  }
+
+  void ShiftRight(unsigned index);
+
+  // Rebalance a full child at position pos, at which we tried to insert at insert_pos.
+  // Returns the node and the position to insert into if rebalancing successful.
+  // Returns nullptr if rebalancing did not succeed.
+  std::pair<BTreeNode*, unsigned> RebalanceChild(unsigned pos, unsigned insert_pos);
+
+  // Inserts item into a leaf node.
+  // Assumes: the node is IsLeaf() and has some space.
+  void LeafInsert(unsigned index, Key_t item) {
+    assert(IsLeaf() && NumItems() < MaxItems());
+    InsertItem(index, item);
+  }
+
+  void InnerInsert(unsigned index, Key_t item, BTreeNode* child) {
+    InsertItem(index, item);
+    SetChild(index + 1, child);
+  }
+
+  // Tries to merge the child at position pos with its sibling.
+  // If we did not succeed to merge, we try to rebalance.
+  // Returns retired BTreeNode* if children got merged and this parent node's children
+  // count decreased, otherwise, we return nullptr (rebalanced).
+  BTreeNode* MergeOrRebalanceChild(unsigned pos);
+
+  void Validate(Key_t upper_bound) const;
+
+ private:
+  void RebalanceChildToLeft(unsigned child_pos, unsigned count);
+  void RebalanceChildToRight(unsigned child_pos, unsigned count);
+
+  void MergeFromRight(Key_t key, BTreeNode* right);
+
+  void InsertItem(unsigned index, Key_t item) {
+    assert(index <= num_items_);
+    assert(index == 0 || Key(index - 1) < item);
+    assert(index == num_items_ || Key(index) > item);
+
+    ShiftRight(index);
+    SetKey(index, item);
+  }
+
+  uint8_t* KeyPtr(unsigned index) {
+    uint8_t* ptr = reinterpret_cast<uint8_t*>(this) + kKeyOffset + kKeySize * index;
```

Comment:
How does this work? This class has no members (did you mean for the anonymous `union` to be a member?), and so here you return a pointer outside the memory of this class.
Do you assume that whenever this is used, it will be directly superseded by another buffer? Why not have the buffer internally?

Reply:
added comment.